lib/AST/ASTDemangler.cpp

```diff
@@ -47,63 +47,117 @@ Type swift::Demangle::getTypeForMangling(ASTContext &ctx,
 
 
 Type
-ASTBuilder::createBuiltinType(const std::string &mangledName) {
-  // TODO
+ASTBuilder::createBuiltinType(StringRef builtinName,
+                              StringRef mangledName) {
+  if (builtinName.startswith(BUILTIN_TYPE_NAME_PREFIX)) {
+    SmallVector<ValueDecl *, 1> decls;
+
+    ModuleDecl::AccessPathTy accessPath;
+    StringRef strippedName =
+          builtinName.drop_front(strlen(BUILTIN_TYPE_NAME_PREFIX));
+    Ctx.TheBuiltinModule->lookupValue(accessPath,
+                                      Ctx.getIdentifier(strippedName),
+                                      NLKind::QualifiedLookup,
+                                      decls);
+    
+    if (decls.size() == 1 && isa<TypeDecl>(decls[0]))
+      return cast<TypeDecl>(decls[0])->getDeclaredInterfaceType();
+  }
+
   return Type();
 }
 
-NominalTypeDecl *
-ASTBuilder::createNominalTypeDecl(StringRef mangledName) {
+GenericTypeDecl *ASTBuilder::createTypeDecl(StringRef mangledName,
+                                            bool &typeAlias) {
   Demangle::Demangler Dem;
   Demangle::NodePointer node = Dem.demangleType(mangledName);
   if (!node) return nullptr;
 
-  return createNominalTypeDecl(node);
+  return createTypeDecl(node, typeAlias);
 }
 
 ProtocolDecl *
 ASTBuilder::createProtocolDecl(const Demangle::NodePointer &node) {
-  return dyn_cast_or_null<ProtocolDecl>(createNominalTypeDecl(node));
+  bool typeAlias;
+  return dyn_cast_or_null<ProtocolDecl>(
+    createTypeDecl(node, typeAlias));
 }
 
-Type ASTBuilder::createNominalType(NominalTypeDecl *decl) {
+Type ASTBuilder::createNominalType(GenericTypeDecl *decl) {
+  auto *nominalDecl = dyn_cast<NominalTypeDecl>(decl);
+  if (!nominalDecl)
+    return Type();
+
   // If the declaration is generic, fail.
-  if (decl->isGenericContext())
+  if (nominalDecl->isGenericContext())
     return Type();
 
-  return decl->getDeclaredType();
+  return nominalDecl->getDeclaredType();
```

Comment:
`getDeclaredInterfaceType()` because it's preferred, even though it's the same due to the `isGenericContext()` check above?